### ISSUE TITLE
^Z/^Y-undo/re-do shortcuts stop working when Menu is hidden (F10)

### ISSUE THREAD

**Issue by tjavdar** [xournalpp/xournalpp#2082]
**Affects versions :**
 - OS: Archlinux
 - uname -a: Linux xe 5.7.4-arch1-1, x86_64 GNU/Linux
 - Desktop environment: i3
 - Which version of libgtk:  1:3.24.20-1
 - Version of Xournal++: 1.0.18, Git commit: 535d261d
 - Installation method:  Arch repo

**Describe the bug**
The shortcuts
       Ctrl-Z (undo last stroke) and Ctrl-Y (redo undone stroke)
stop working after the menu is hidden via F10.

**To Reproduce**
1. Start xournal and make a stroke. Press ^Z to undo it and ^Y to redo it.
2. Press F10 to hide the "File  Edit View ..." top menu.
3. Pressing ^Z and ^Y stop working

**Expected behavior**
 Ctrl-Z (undo) and Ctrl-Y (redo) shortcuts should work no matter if the Menu is on-screen or not

**Comment by s1-ranjan** [comment]
I would like to add one more important observation in this most of the shortcut keys are non functional when menu is hidden.

**Comment by duncanawoods** [comment]
I'll take a look at this. I believe it is a default GTK behaviour which can be fixed by binding the accelerators to the main window, not the menu items:

https://stackoverflow.com/questions/19707775/accelerators-stop-responding-when-menubar-gets-hidden

**Comment by edo0** [comment]
Duplicate of  #1002

**Comment by duncanawoods** [comment]
I've taken a look at this. 

I've tried to iterate through the menu items use them to generate global shortcuts but my attempts to invoke the bound GClosure haven't worked and the menu item widget itself isn't valid when the menu is hidden. I don't think there is an alternative to manually creating the bindings duplicating/replacing the menu accelerators. 

We can't have both the global shortcuts and the menu accelerators bound to the same keys at the same time.

There is an argument for removing the menu accelerators. To match the current UI we would then have to manually add labels to indicate the global keys.

For now, I have implemented a way to enable/disable global accelerators when the menubar is shown/hidden and it works but I would like some confirmation to proceed before I implement the rest of the accelerators. It uses an intermediary representation of the accelerators which will be a good beginning for key maps #919 where most of the work will be the settings pages to configure the shortcuts.

```cpp
typedef void(*AccelCallback)(MainWindow* mainWindow);

struct GlobalAccelerator {
    AccelCallback callback; 
    guint key; 
    GdkModifierType modifierType; 
    GtkAccelFlags accelFlags;
};

class MainWindow {
private:
    void initGlobalAccelerators();
    void connectGlobalAccelerators();
    void disconnectGlobalAccelerators();

    GtkAccelGroup* globalAccelGroup;
    vector<GlobalAccelerator> globalAccelerators;
}
```

And the binding:

```cpp
void MainWindow::initGlobalAccelerators() {
    this->globalAccelGroup = gtk_accel_group_new();
    gtk_window_add_accel_group(GTK_WINDOW(this->window), this->globalAccelGroup);

    this->globalAccelerators = {
        { toggleToolbarVisible, GDK_KEY_F9, GdkModifierType(0), GtkAccelFlags(0) }
    };

    Settings* settings = control->getSettings();
    if (!settings->isMenubarVisible()) {
        this->connectGlobalAccelerators();
    }
}

void MainWindow::connectGlobalAccelerators() {
    for (auto &accel : this->globalAccelerators) {
        gtk_accel_group_connect(this->globalAccelGroup, accel.key, accel.modifierType, accel.accelFlags,
                                g_cclosure_new_swap (G_CALLBACK (accel.callback), this, NULL));
    }
}

void MainWindow::disconnectGlobalAccelerators() {
    for (auto &accel : this->globalAccelerators) {
        gtk_accel_group_disconnect_key(this->globalAccelGroup, accel.key, accel.modifierType);
    }
}
```

**Comment by rolandlo** [comment]
@duncanawoods Can you please provide a version for testing purpose?

**Comment by rolandlo** [comment]
This has been solved in #2324.
